doc/adrs/173_support_pagination_in_rest_apis.adoc

```diff
@@ -0,0 +1,51 @@
+= [ADR-173] Support Pagination in REST APIs
+
+== Context
+
+REST APIs may return a huge amount of data in Sirius Web.
+
+=== Current behavior
+
+The REST APIs don't support pagination options in Sirius Web.
+
+== Decision
+
+The decision is to support pagination in all appropriate GET REST APIs described in the Systems Modeling Application Programming Interface (API) and Services section "8.1 REST/HTTP PSM" coming from https://www.omg.org/spec/SystemsModelingAPI/.
+
+The following text has been extracted form the specification:
+
+```
+The REST/HTTP PSM uses a Cursor-based pagination strategy for the responses received from the GET requests.
+The following 3 query parameters can be specified in any GET request that returns a collection of records.
+1. `page[size]` specifies the maximum number of records that will be returned per page in the response
+2. `page[before]` specifies the URL of the page succeeding the page being requested
+3. and `page[after]` specifies the URL of a page preceding the page being requested
+If neither `page[before]` nor `page[after]` is specified, the first page is returned with the same number of records as specified in the `page[size]` query parameter.
+If the `page[size]` parameter is not specified, then a default page size is used, which can be set by the API provider.
+The `Link` header in the response includes links (URLs) to the previous page and the next page, if any, for the given page in the response.
+The specification of these links is conformant to the https://datatracker.ietf.org/doc/html/rfc5988[IETF Web Linking standard].
+As an example, the value of the `Link` response header is shown below.
+The `rel` value associated with each page link specifies the type of relationship the linked  page has with the page returned in the response.
+Page link specified with `rel` value as `next` is the link for the next (or succeeding) page to the page returned in the response, and the page link specified with `rel` value as `prev` is the link for the previous (or preceding) page to the page returned in the response.
+
+Examples:
+
+<http://sysml2-api-host:9000/projects?
+page[after]=MTYxODg2MTQ5NjYzMnwyMDEwOWY0MC00ODI1LTQxNmEtODZmNi03NTA4YWM0MmEwMjE&
+page[size]=3>; rel="next",
+<http://sysml2-api-host:9000/projects?
+page[before]=MTYxODg2MTQ5NjYzMnwxMDg2MDFjMS1iNzk1LTRkMGEtYTFiYy1lZjEyYmMwNTU5ZjI&
+page[size]=3>; rel="prev"
+```
+
+However the https://docs.spring.io/spring-data/rest/reference/data-commons/repositories/core-extensions.html#core.web.basic.paging-and-sorting[pagination in spring data] does not looks exactly the same than the one proposed by the SystemsModelingAPI specification.
```

Comment:
If I'm not mistaken, [here](https://docs.spring.io/spring-data/rest/reference/data-commons/repositories/scrolling.html#page-title) is the API that should be used. You can see it in action [here](https://www.baeldung.com/spring-data-jpa-scroll-api) for example.

I think we will need to use it with keyset filtering to ask, for example, for the next 10 projects after the one with the identifier XXX.

Reply:
This will need to be prototyped on projects first since they are the perfect candidate for that. We should try to remove support for page-based APIs in favor of scroll based ones entirely not support both to have something like the example below. It was always the plan to make it disappear in the GraphQL API to follow [Relay best practices](https://relay.dev/graphql/connections.htm) that are supported by GraphQL Java:

```
public interface IProjectSearchService {
    boolean existsById(UUID projectId);
    Optional<Project> findById(UUID projectId);
    Window<Project> findAll(KeyScrollPosition position, int limit);
}
```

You could start prototyping with this API using some integration test on the backend to see if it allows you to retrieve what you want.

Reply:
With something like this:

```
KeyScrollPosition position = ScrollPosition.of(
    Map.of("id", "MTYxODg2MTQ5NjYzMnwyMDEwOWY0MC00ODI1LTQxNmEtODZmNi03NTA4YWM0MmEwMjE"),
    Direction.FORWARD
);

var projectWindow = this.projectSearchService.findAll(position, 3);
```